File: Algorithms/Medium/143.Reorder-List/ReorderList.go
package medium

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time complexity: O(n)
// Space complexity: O(n)
func reorderList_1(head *ListNode) {
	slow, fast := head, head
	stack := make([]*ListNode, 0)
	pop := func() *ListNode {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		return top
	}

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	end := slow
	for slow != nil {
		stack = append(stack, slow)
		slow = slow.Next
	}

	for head != end {
		nxt := head.Next
		head.Next = pop()
		head.Next.Next = nxt
		head = nxt
	}

	end.Next = nil
}

// Time complexity: O(n)
// Space complexity: O(1)
func reorderList_2(head *ListNode) {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	var prev *ListNode
	curr := slow.Next
	slow.Next = nil
	for curr != nil {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}

	first, curr := head, prev
	for curr != nil {
		tmp1, tmp2 := first.Next, curr.Next
		first.Next = curr
		curr.Next = tmp1
		first, curr = tmp1, tmp2
	}
}

File: Algorithms/Medium/143.Reorder-List/ReorderList.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        stack = []
        slow, fast = head, head

        # getting the middle of the linked list
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        end = slow
        while slow:
            stack.append(slow)
            slow = slow.next

        while head != end:
            nxt = head.next
            node = stack.pop()
            head.next = node
            node.next = nxt
            head = head.next.next

        end.next = None